### PR TITLE
[DAG] Fold extract_subvector(insert_subvector(x,y,c1),c2) --> extract_subvector(y,c2-c1)

### PR DESCRIPTION
**Pull request by RKSimon**
If the extract_subvector is cheap, attempt to extract directly from an inserted subvector

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 1]
```diff
@@ -24459,6 +24459,23 @@ SDValue DAGCombiner::visitEXTRACT_SUBVECTOR(SDNode *N) {
       if (!LegalOperations || TLI.isOperationLegal(ISD::SPLAT_VECTOR, NVT))
         return DAG.getSplatVector(NVT, SDLoc(N), V.getOperand(0));
 
+  // extract_subvector(insert_subvector(x,y,c1),c2)
+  //  --> extract_subvector(y,c2-c1)
```
Comment (phoebewang) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:24463]: The `c1`, `c2` are indices based on type of `y` and `NVT`, we cannot sub them directly given they have difference scale.
For example, if we insert `v16i8` to `v32i8` at index 1, then extra `v8i8` at index 2.
```
| v16i8(y)|  v16i8  |
|v8i8| NVT|v8i8|v8i8|
```
The new index is `c2 - c1 * (NumInsElts/NumSubElts) = 0` instead of `c2 - c1 = 1`.

Comment (RKSimon) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:24463]: Aren't you confusing the AVX subvector instructions (which use subvector offset indices) with the DAG subvector nodes (which use element offset indices)? See getInsertVINSERTImmediate

**Comment by RKSimon** [comment]
ping?

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 1 above)
Comment (phoebewang) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:24463]: Thanks! I misunderstood it.

**Review by phoebewang — APPROVED** [review]
LGTM.

**Comment by steven-johnson** [comment]
Echoing a comment from a specific commit:

Just a heads up, looks like this broke Halide SVE2 tests here: https://github.com/halide/Halide/blob/main/test/correctness/simd_op_check_sve2.cpp#L585.

Assertion triggered:

llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:7061 in SDValue llvm::SelectionDAG::getNode(unsigned int, const SDLoc &, EVT, SDValue, SDValue, const SDNodeFlags): (VT.isFixedLengthVector() || N1VT.isScalableVector()) && "Cannot extract a scalable vector from a fixed length vector!"

I'm investigating now to see if this is a bug on the Halide side or the LLVM side, will update here as I get more info.

**Comment by steven-johnson** [comment]
I think this is a subtle bug in the changed code here -- I think the code above doesn't consider the possibility of scalable / fixed vector combinations, and Halide's tests (which tend to exercise a lot of corner cases in LLVM vector support) are breaking this. I'd like to request this get rolled back (to keep Halide unbroken) until we can work out exactly why this is breaking in this fashion, and hopefully re-land it after doing testing with Halide.

TL;DR: this needs to be rolled back.

**Comment by steven-johnson** [comment]
Enclosed is an IR file that will replicate the crash when run through llc.
[test_op_frecpe_float32_x8_0.zip](https://github.com/llvm/llvm-project/files/14984651/test_op_frecpe_float32_x8_0.zip)

**Comment by RKSimon** [comment]
Thanks @antmox - got a repro now
